src/ports/postgres/modules/graph/pagerank.py_in

```diff
@@ -0,0 +1,290 @@
+        CREATE TEMP TABLE {edge_temp_table} AS
+        SELECT * FROM {edge_table}
+        {distribution}
+        """.format(**locals()))
+    # GPDB and HAWQ have distributed by clauses to help them with indexing.
+    # For Postgres we add the indices manually.
+    sql_index = m4_ifdef(<!__POSTGRESQL__!>,
+        <!"""CREATE INDEX ON {edge_temp_table} ({src});
+        """.format(**locals())!>,
+        <!''!>)
+    plpy.execute(sql_index)
+
+    nvertices = plpy.execute("""
+            SELECT COUNT({0}) AS cnt
+            FROM {1}
+        """.format(vertex_id, vertex_table))[0]["cnt"]
+    init_value = 1.0/nvertices
+    random_prob = (1.0-damping_factor)/nvertices
+    cur = unique_string(desp='cur')
+    message = unique_string(desp='message')
+    plpy.execute("""
+            CREATE TEMP TABLE {cur} AS
+            SELECT {vertex_id}, {init_value}::DOUBLE PRECISION AS pagerank
+            FROM {vertex_table}
+        """.format(**locals()))
+    v1 = unique_string(desp='v1')
+
+    out_cnts = unique_string(desp='out_cnts')
+    out_cnts_cnt = unique_string(desp='cnt')
+    # Compute the out-degree of every node in the graph.
+    cnts_distribution = m4_ifdef(<!__POSTGRESQL__!>, <!''!>,
+        <!"DISTRIBUTED BY ({0})".format(vertex_id)!>)
+
+    plpy.execute("""
+        DROP TABLE IF EXISTS {out_cnts};
+        CREATE TEMP TABLE {out_cnts} AS
+        SELECT {src} AS {vertex_id}, COUNT({dest}) AS {out_cnts_cnt}
+        FROM {edge_table}
+        GROUP BY {src}
+        {cnts_distribution}
+        """.format(**locals()))
+
+    for i in range(max_iter):
+        plpy.execute("""
+                CREATE TABLE {message} AS
+                SELECT {edge_temp_table}.{dest} AS {vertex_id},
+                        SUM({v1}.pagerank/{out_cnts}.{out_cnts_cnt})*{damping_factor}+{random_prob} AS pagerank
+                FROM {cur}, {out_cnts}, {edge_temp_table}, {cur} AS {v1}
+                WHERE {edge_temp_table}.{dest}={cur}.{vertex_id} AND
+                      {out_cnts}.{vertex_id}={edge_temp_table}.{src} AND
+                      {v1}.{vertex_id}={edge_temp_table}.{src}
+                GROUP BY {edge_temp_table}.{dest}
+            """.format(**locals()))
+        # If there are nodes that have no incoming edges, they are not captured in the message table.
+        # Insert entries for such nodes, with random_prob.
+        plpy.execute("""
+                INSERT INTO {message}
+                SELECT {vertex_id}, {random_prob}::DOUBLE PRECISION AS pagerank
+                FROM {cur}
+                WHERE {vertex_id} NOT IN (
+                    SELECT {vertex_id}
+                    FROM {message}
+                )
+            """.format(**locals()))
+        # Check for convergence
+        if 'e' in str(threshold):
+            thresh = abs(int(str(threshold).split('e')[-1]))
+        else:
+            thresh = len(str(threshold).split('.')[1])
+        if plpy.execute("""
+                SELECT (
```

Comment:
Will take care of these convergence test comments as part of https://issues.apache.org/jira/browse/MADLIB-1082.